File: Roboti/app/src/main/java/cz/muni/fi/umimecesky/activity/TrainingActivity.java
package cz.muni.fi.umimecesky.activity;

import android.os.Build;
import android.os.Bundle;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import com.google.gson.Gson;

import java.util.List;

import cz.muni.fi.umimecesky.R;
import cz.muni.fi.umimecesky.pojo.Category;
import cz.muni.fi.umimecesky.pojo.FillWord;
import cz.muni.fi.umimecesky.utils.Utils;

import static cz.muni.fi.umimecesky.utils.Constant.LAST_FILLED_WORD;
import static cz.muni.fi.umimecesky.utils.Constant.TICKED_CATEGORIES_EXTRA;
import static cz.muni.fi.umimecesky.utils.Constant.TRAINING_NEW_WORD_DELAY;

public class TrainingActivity extends BaseAbstractActivity {

    private TextView explanationText;

    @Override
    @SuppressWarnings("unchecked assignment")
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_training);

        setCheckedCategories((List<Category>) getIntent().getSerializableExtra(TICKED_CATEGORIES_EXTRA));

        setWordText((TextView) findViewById(R.id.word));
        setCategoryText((TextView) findViewById(R.id.category));
        explanationText = (TextView) findViewById(R.id.explanationText);
        setVariant1((Button) findViewById(R.id.firstButton));
        setVariant2((Button) findViewById(R.id.secondButton));
        super.init();

        setLastUsedWord();


        View.OnTouchListener touchListener = new View.OnTouchListener() {
            @Override
            public boolean onTouch(View v, MotionEvent event) {
                switch (event.getAction()) {
                    case MotionEvent.ACTION_DOWN:
                        v.setAlpha(.8f);
                        return false;
                    case MotionEvent.ACTION_UP:
                        // RELEASED
                        v.setAlpha(1);
                        return false;
                }
                return false;
            }
        };

        getVariant1().setOnTouchListener(touchListener);
        getVariant2().setOnTouchListener(touchListener);

    }

    @Override
    protected void chosenWrong(Button button) {
        setWrong(button);
        showExplanation();
    }

    @Override
    protected void chosenCorrect(Button button) {
        setCorrect(button);

        button.postDelayed(new Runnable() {
            @Override
            public void run() {
                setButtonsEnabled();
                setNewRandomWord();
            }
        }, TRAINING_NEW_WORD_DELAY);

    }

    private void showExplanation() {
        String explanation = getCurrentWord().getExplanation();
        if (!explanation.isEmpty()) {
            explanationText.setText(explanation);
            explanationText.setVisibility(View.VISIBLE);
        }
    }

    private void hideExplanation() {
        explanationText.setVisibility(View.INVISIBLE);
    }

    private void setLastUsedWord() {
        String json = getSharedPref().getString(LAST_FILLED_WORD, null);
        FillWord lastWord = new Gson().fromJson(json, FillWord.class);

        if (lastWord != null) {
            setWord(lastWord);
        } else {
            setNewRandomWord();
        }
    }

    private void setNewRandomWord() {
        FillWord word;
        if (getCheckedCategories() == null || getCheckedCategories().isEmpty()) {
            word = getWordHelper().getRandomFilledWord();
        } else {
            List<Integer> categoryIDs = Utils.convertCategoriesToIDs(getCheckedCategories());
            word = getWordCategoryHelper().getRandomCategoryWord(categoryIDs);
        }
        Log.v("random word", String.valueOf(word));
        setWord(word);
    }

    @Override
    protected void setWord(FillWord word) {
        super.setWord(word);

        hideExplanation();
        setCategoryName();
        setButtonsEnabled();
    }

    private void setCategoryName() {
        int categoryId = getWordCategoryHelper().getCategoryId(getCurrentWord().getId());
        Category category = getCategoryHelper().findCategory(categoryId);
        if (category != null) getCategoryText().setText(category.getName());
    }

    @Override
    protected void onPause() {
        super.onPause();
        String json = new Gson().toJson(getCurrentWord());
        getSharedPref().edit().putString(LAST_FILLED_WORD, json).apply();
    }

    @Override
    protected void onResume() {
        super.onResume();
        int flags = View.SYSTEM_UI_FLAG_HIDE_NAVIGATION; // hides nav bar (buttons)
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.KITKAT) {
            flags |= View.SYSTEM_UI_FLAG_IMMERSIVE_STICKY;
        }
        getWindow().getDecorView().setSystemUiVisibility(flags);
    }

}

File: Roboti/app/src/main/java/cz/muni/fi/umimecesky/activity/TrainingActivity.kt
package cz.muni.fi.umimecesky.activity

import android.content.pm.ActivityInfo
import android.os.Bundle
import android.view.MotionEvent
import android.view.View
import android.widget.Button
import com.google.firebase.analytics.FirebaseAnalytics
import cz.muni.fi.umimecesky.R
import cz.muni.fi.umimecesky.db.helper.joinCategoryWordOpenHelper
import cz.muni.fi.umimecesky.labyrinth.Constant.isTablet
import cz.muni.fi.umimecesky.pojo.Category
import cz.muni.fi.umimecesky.pojo.FillWord
import cz.muni.fi.umimecesky.utils.Constant.TICKED_CATEGORIES
import cz.muni.fi.umimecesky.utils.Constant.TRAINING_NEW_WORD_DELAY_MS
import cz.muni.fi.umimecesky.utils.TrainingProgressBar
import kotlinx.android.synthetic.main.activity_training.categoryText
import kotlinx.android.synthetic.main.activity_training.explanationText
import kotlinx.android.synthetic.main.activity_training.firstButton
import kotlinx.android.synthetic.main.activity_training.secondButton
import kotlinx.android.synthetic.main.activity_training.seriesProgressBar
import kotlinx.android.synthetic.main.activity_training.word
import org.jetbrains.anko.configuration
import org.jetbrains.anko.landscape

class TrainingActivity : BaseAbstractActivity() {

    companion object {
        private const val LAST_SEEN_WORD = "lastSeenWord"
    }

    private lateinit var trainingProgressBar: TrainingProgressBar

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_training)
        super.initUI(word, firstButton, secondButton)

        // TODO: make XML on small displays in landscape mode durable
        if (configuration.landscape && !isTablet())
            requestedOrientation = ActivityInfo.SCREEN_ORIENTATION_PORTRAIT

        @Suppress("UNCHECKED_CAST")
        checkedCategories = intent.getSerializableExtra(TICKED_CATEGORIES) as List<Category>
        logCheckedCategories()

        setNewRandomWord()

        trainingProgressBar = TrainingProgressBar(this, seriesProgressBar)

        // TODO: show better when button is pressed
        val touchListener = View.OnTouchListener { v, event ->
            when (event.action) {
                MotionEvent.ACTION_DOWN -> {
                    v.alpha = .8f
                    return@OnTouchListener false
                }
                MotionEvent.ACTION_UP -> {
                    // RELEASED
                    v.alpha = 1f
                    return@OnTouchListener false
                }
            }
            false
        }

        variant1.setOnTouchListener(touchListener)
        variant2.setOnTouchListener(touchListener)

    }

    private fun logCheckedCategories() {
        val bundle = Bundle()
        bundle.putString(FirebaseAnalytics.Param.CONTENT_TYPE, checkedCategories.toString())
        FirebaseAnalytics.getInstance(this).logEvent("checked_categories", bundle)
    }

    override fun onSaveInstanceState(outState: Bundle?) {
        outState?.putSerializable(LAST_SEEN_WORD, currentWord)
        super.onSaveInstanceState(outState)
    }

    override fun onRestoreInstanceState(savedInstanceState: Bundle?) {
        super.onRestoreInstanceState(savedInstanceState)

        if (savedInstanceState != null) {
            setWord(savedInstanceState.get(LAST_SEEN_WORD) as FillWord)
        } else {
            setNewRandomWord()
        }
    }

    override fun chosenWrong(button: Button) {
        setWrong(button)
        showExplanation()
        trainingProgressBar.incrementIncorrectAttemptsCounter()
    }

    override fun chosenCorrect(button: Button) {
        setCorrect(button)
        trainingProgressBar.processCorrectResult()

        button.postDelayed({
            enableButtons()
            setNewRandomWord()
        }, TRAINING_NEW_WORD_DELAY_MS)

    }

    private fun showExplanation() {
        val explanation = currentWord.explanation
        if (!explanation.isEmpty()) {
            explanationText.text = explanation
            explanationText.visibility = View.VISIBLE
        }
    }

    private fun hideExplanation() {
        explanationText.visibility = View.INVISIBLE
    }

    private fun setNewRandomWord() {
        try {
            setWord(joinCategoryWordOpenHelper.getRandomCategoryWord(checkedCategories))
        } catch (e: IllegalArgumentException) {
            // TODO: remove later, just for better exception info
            val bundle = Bundle()
            bundle.putString("exception_message", e.message)
            bundle.putString("which_categories", checkedCategories.toString())
            FirebaseAnalytics.getInstance(this).logEvent("db_problem", bundle)
        }
    }

    override fun setWord(word: FillWord?) {
        super.setWord(word)

        hideExplanation()
        setCategoryName()
        enableButtons()
    }

    private fun setCategoryName() {
        categoryText.text = joinCategoryWordOpenHelper.getCategory(currentWord).name
    }
}
